### PR TITLE
0089/stats

### PR DESCRIPTION
**Pull request by davbyr**
EDIT: There is a lot of output during unit testing, which I'm not sure how to switch off.

@jpolton @anwiseNOCL Sorry about how large this branch is, I got a bit carried away and all the changes are kind of related. Nothing I have done should affect either of your work. This is getting near ready for review, I'm just trying to figure out why Github is failing the run the branch. Hopefully I haven't broken anything significant..

**CRPS**
CRPS is no longer its own object, but instead a small module with some related methods inside. To obtain CRPS, you now do it via the observation object, i.e. TIDEGAUGE or ALTIMETRY. Calling .crps() will either return a new copy of the original object containing just CRPS related data or save the new variables to the original object. This depends on whether create_new_object = True in .crps(). Having it live inside an object of the same type it originated means it can access all the same plotting routines as the observations, which is useful. EXAMPLE:

```
# Load altimetry and NEMO objects. Calculate CRPS between a variable in each.
altimetry = coast.ALTIMETRY(file)
nemo = coast.NEMO(file)
crps = altimetry.crps(nemo, model_var_str = 'ssh', obs_var_str = 'sla_filtered', create_new_object=True)
# Take a look
crps.quick_plot('crps')
```

**Obs_operator**
NEMO objects can now be interpolate to both ALTIMETRY and TIDEGAUGE object spaces. EXAMPLE

```
# Interpolate NEMO object to tide gauge location and times + load tidegauge
tg = coast.TIDEGAUGE(file)
# Interpolate ssh to tg
tg.obs_operator(nemo, 'ssh')
# interp_ssh will now sit inside tg object
```

**TIDEGAUGE**
A TIDEGAUGE object now contains data for a single tidegauge. It contains methods for reading from GESLA files. It also contains static methods for reading multiple tide gauge files (which are stored in a list of TIDEGAUGE objects) and for plotting using this list if desired (so plotting up multiple tide gauge locations). Example for single gauge above. For multiple:

```
# Load multiple gauge files using wildcard
tg_list = coast.TIDEGAUGE.create_multiple_tidegauge('./*', date_start = date0, date_end = date1)
# Plot up their locations
coast.TIDEGAUGE.plot_on_map_multiple(tg_list)
```

**STATS**
STATS is not an object but, similar to CRPS, is called via the observation objects. There are some smaller methods which calculate individual basic statistics (error, AE, MAE, RMSE, correlations etc) which return single values or DataArrays. .basic_stats will call all of these and compile the output into a new xarray dataset. This can either then be merged with the original object or create a new object, similar to the CRPS method. EXAMPLE:

```
# Calculate basic statistics for tidegauge and interpolate model ssh. Save to new object:
stats = tg.basic_stats('interp_ssh', 'sea_level', create_new_object = True)
# This contains Errors, Absolute errors, MAE, RMSE, correlations and covariance.
# Or get a monthly mean sea level for the first week of Jan:
mmean = tg.time_mean('sea_level', date0 = '2007-01-01', date1 = '2007-01-07')
```

**Organisation**
Three new utility modules: general_util, stats_util, plot_util. All quite empty at the moment but contain essentially static methods that are widely used.

**Unit testing**
Updated my parts to now be sepcific to TIDEGAUGE and ALTIMETRY objects. Also added a unit template to the top of the file:

```
Unit template:
    
'''
#-----------------------------------------------------------------------------#
# ( ## ) Subsection title                                                     #
#                                                                             #
'''
subsec = subsec+1
# <Introduction>

try:
    # Do a thing
    
    #TEST: <description here>
    check1 = #<Boolean>
    check2 = #<Boolean> 
    if check1 and check2:
        print(str(sec) + chr(subsec) + " OK - ")
    else:
        print(str(sec) + chr(subsec) + " X - ")
    
except:
    print(str(sec) + chr(subsec) +' FAILED.')

```

**Cleaning Up**
Cleaned up some of my docstrings and updated some others. Removed *some* of my redundant methods and files, most notably interpolate_along_dimension.

### CODE REVIEW

File: unit_testing/unit_test.py [hunk 1]
```diff
@@ -14,7 +51,15 @@
 import xarray as xr
 import matplotlib.pyplot as plt
 import datetime
+import os.path as path
 import logging
+from . import general_utils
```
Comment (jpolton) [inline unit_testing/unit_test.py:56]: This breaks if you are not in ``COAsT/coast``. Is there a way to use ``os.path`` to sort this? Not sure. Not a big issue.

Comment (davbyr) [reply unit_testing/unit_test.py:56]: Maybe putting it into __init__.py like the object files. I'll look into it

File: unit_testing/unit_test.py [hunk 2]
```diff
@@ -671,70 +717,70 @@
...
-# ( 6b ) CRPS Map Plots                                                       #
+# ( 6b ) Altimetry subsetting                                                 #
 #                                                                             #
+'''
 subsec = subsec+1
-plt.close('all')
+# The altimetry that we loaded is global so lets subset it for the North
+# West European Shelf.
 try:
-    fig, ax = crps.map_plot()
```
Comment (jpolton) [inline unit_testing/unit_test.py:712]: The CRPS map plot disappeared

File: unit_testing/unit_test.py [hunk 3]
```diff
@@ -671,70 +717,70 @@
...
-#                                                                             #
+    ind = altimetry.subset_indices_lonlat_box([-10,10], [45,60])
+    ind = ind[::4]
+    altimetry_nwes = altimetry.isel(t_dim=ind) #nwes = northwest europe shelf
 
-plt.close('all')
-subsec = subsec+1
-try:
-    fig, ax = crps.cdf_plot(0)
-    fig.savefig(dn_fig + 'crps_cdf_plot.png')
```
Comment (jpolton) [inline unit_testing/unit_test.py:727]: The CDF plot disappeared

**Review by jpolton — COMMENTED** [review]
Minor typos in unit testing. Pull request on its way.
7e fails (multiple gauge load) because I have 9 objects not too:
```
date0 = datetime.datetime(2007,1,10)
date1 = datetime.datetime(2007,1,12)
tidegauge_list = coast.TIDEGAUGE.create_multiple_tidegauge('./example_files/tide_gauges/*',date0,date1)

tidegauge_list
Out[6]: 
[<coast.TIDEGAUGE.TIDEGAUGE at 0x7fb745983f90>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb746944d10>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb97da52490>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb739064d10>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb73646be50>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb97d651110>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb7459943d0>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb97cd42750>,
 <coast.TIDEGAUGE.TIDEGAUGE at 0x7fb744e4f2d0>]

len(tidegauge_list)
9
```

File: unit_testing/unit_test.py (hunk 3 above)
Comment (davbyr) [reply unit_testing/unit_test.py:727]: CRPS has been moved into observation objects, so their plotting routines can be used to visualise it. CDF plot's machinery still exists but I've removed the method for now (thinking of a better way to implement it).

**Comment by davbyr** [comment]
@jpolton I should have said. I have changed the testing to just two tide gauge files to save on run time (lowestoft and llandudno). So we will need to update the ERDAP server

**Comment by jpolton** [comment]
On the second ``run unit_testing/unit_test.py`` do you get loads of extra messages in the console?

**Comment by davbyr** [comment]
Yes I do, they're coming from the logging stuff somewhere, but I'm not sure why..

**Comment by anwiseNOCL** [comment]
@davbyr I'm getting some errors in unit testing for 6e and 7d. It turns out that the version of sckit-learn I have doesn't support 
the parameter 'square' in 
`rmse = metrics.mean_squared_error(var0, var1, squared=False)` line 278 ALTIMETRY

We will need to update the version of scikit-learn required in the package to 0.22.2 or 0.23 (latest).

**Comment by davbyr** [comment]
@anwiseNOCL You mean the `squared=False`? I can always remove that and square root it using numpy instead. If you mean the metrics object, then I can also just remove that and do it with numpy.

**Comment by jpolton** [comment]
CRPS.py still exits

**Review by anwiseNOCL — APPROVED** [review]
I updated to the latest version of scikit-learn and it works. All tests work for me.

**Comment by davbyr** [comment]
@anwiseNOCL @jpolton I've merged develop into this branch and it all works for me. Do you two want to check the merged unit test one more time before I merge to develop?

**Review by jpolton — APPROVED** [review]
Ran unit_testing. It _all_ works. Horrar!

**Review by anwiseNOCL — APPROVED** [review]

**Pull request merged**